bolt-design-system/bolt
# Config Refactor + Misc Cleanup and Fixes

## Jira
http://vjira2:8080/browse/BDS-1397

## Summary
Refactors and reorganizes globally shared configs, packages up existing files related to Nightwatch and Jest testing, and cleans up a variety of items related to configuration, testing and the build process. 

Note: this also includes 2 fixes to the carousel component found when testing these updates out -- can move to a separate PR if necessary.

## Details
* config-presets —> rename folder to `configs`
* Fixed a few carousel component bugs
    * `maxslides_per_view` -> `max_slides_per_view` typo
    * Removed non-existent `free_mode` prop on gallery demo (schema calls this prop something different)
    * Fixed an issue with the slidesPerGroup config not getting calculated correctly when the component initially boots up
* `browserslist-config`
    * NPM Package renamed to `@bolt/browserslist-config` to match the naming convention recommended by browserslist (allows for easier sharing of the config across projects — https://github.com/browserslist/browserslist#shareable-configs)
    * Note: this shouldn’t technically be a major breaking change since the older package still works
    * @todo: the older NPM package to this new package + add deprecated NPM warning
* `stylelint-config`
    * Similar in updates to browserslist — updated NPM package name to match recommend naming convention
* Autoprefixer / Babel / Browserslist
    * The newer version of Browserslist requires a change in how the config is wired up to other packages like Babel and Autoprefixer. The updates in this PR reflect this update.
* `@bolt/build-utils`
    * Added a new sub-package to @bolt/build-tools package which includes just the utility and helper functions required. This update should help address recursive dependency warnings flagged by Lerna when bootstrapping the project
    * Updated most references in other build tool tasks + external packages using these utils to reference the new package
    * Fixed an issue with the Pattern Lab build task starting to watch for file changes before Webpack finishes the initial build, causing unnecessary recompiles while booting up the Bolt build tools  
    * Updated the Bolt CLI logo shown when booting up the Bolt Build Tools
* General
    * Added new test utility function to allow component tests to check to see if the test is being run online vs being run offline
    * Updated tests to use Puppeteer context method to speed up time required to run puppeteer tests
    * Updated test logic in the Navbar and Autotrack components to modify behavior if the test is being run offline
    * Moved the example SVG added for checking the SVG build process to the top level __tests__ 
    * Auto cleaned up and sorted package.json files updated for more project-wide consistency 
    * Root package.json: removed / relocated dependencies that are config or task-specific vs ones that are required to be kept in the root package.json file.
    * Version bumped a handful of packages pointing to older Bolt dependencies (ex. the new toggle switch component)
    * Fixed an issue with the docs site homepage “double animation” when the page initially loads
    * Minor cleanup of the Travis config file
* Jest config / helper files
    * move majority of related files to a new `config/jest` package
    * Move package dependencies to the new package.json file
* Nightwatch config / helper files
    * move all files to a new `config/nightwatch` package
    * Move package dependencies to the new package.json file
* Added the new `@bolt/testing-helpers` package
    * Consolidates and simplifies common Jest test functionality
    * Reduces the amount of boilerplate code needed within individual tests
    * Moves test-specific dependencies into a single package
    * Speeds up overall time required to run tests

## How to test
1. Confirm the Travis build working as expected
2. Double check these updates here make sense + work locally
3. Any red flags / concerns with the overall changes here?
